### ISSUE TITLE
Implement preventing Sys::Hostname::hostname() from accepting arguments

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17019]
Migrated from [rt.perl.org#134137](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134137) (status was 'pending release')

Searchable as RT134137$

**Comment by p5pRT** [comment]
### From @jkeenan

pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:

\#\#\#\#\#
hostname\(\) doesn't accept any arguments

The function
"hostname\(\)" in the Sys&#8203;::Hostname module has always been
documented to be called with no arguments\. Historically it
has not enforced this\, and has actually accepted and
ignored any arguments\. As a result\, some users have got the
mistaken impression that an argument does something useful\.
To avoid these bugs\, the function is being made strict\.
Passing arguments was deprecated in Perl 5\.28\, and will
become fatal in Perl 5\.32\.
\#\#\#\#\#

This entry was recorded in the following commit&#8203;:

\#\#\#\#\#
commit 0c9c439d08a65206d442724bcd9fb29fa5a7f937
Author&#8203;: Zefram \<zefram@&#8203;fysh\.org>
Date&#8203;:   Wed Dec 6 16&#8203;:28&#8203;:18 2017 \+0000

document hostname\(@&#8203;args\) deprecation

Passing arguments to Sys&#8203;::Hostname&#8203;::hostname\(\) is already
deprecated\. Note the deprecation's agreed removal version
of Perl 5\.32\. Also document it in perldeprecation\.pod\.
\#\#\#\#\#

Make it so\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

Summary of my perl5 \(revision 5 version 31 subversion 0\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 58f4626762668e1c1948832073998af84b2c85d0
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux
&nbsp;    osvers=4\.15\.0\-50\-generic
&nbsp;    archname=x86\_64\-linux
&nbsp;    uname='linux zareason 4\.15\.0\-50\-generic \#54\-ubuntu smp mon may 6 18&#8203;:46&#8203;:08 utc 2019 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-des \-Dusedevel \-Dusemymalloc=y'
&nbsp;    hint=recommended
&nbsp;    useposix=true
&nbsp;    d\_sigaction=define
&nbsp;    useithreads=undef
&nbsp;    usemultiplicity=undef
&nbsp;    use64bitint=define
&nbsp;    use64bitall=define
&nbsp;    uselongdouble=undef
&nbsp;    usemymalloc=y
&nbsp;    default\_inc\_excludes\_dot=define
&nbsp;    bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'
&nbsp;    ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    optimize='\-O2'
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''
&nbsp;    gccversion='7\.4\.0'
&nbsp;    gccosandvers=''
&nbsp;    intsize=4
&nbsp;    longsize=8
&nbsp;    ptrsize=8
&nbsp;    doublesize=8
&nbsp;    byteorder=12345678
&nbsp;    doublekind=3
&nbsp;    d\_longlong=define
&nbsp;    longlongsize=8
&nbsp;    d\_longdbl=define
&nbsp;    longdblsize=16
&nbsp;    longdblkind=3
&nbsp;    ivtype='long'
&nbsp;    ivsize=8
&nbsp;    nvtype='double'
&nbsp;    nvsize=8
&nbsp;    Off\_t='off\_t'
&nbsp;    lseeksize=8
&nbsp;    alignbytes=8
&nbsp;    prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'
&nbsp;    ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/7/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /lib64 /usr/lib64
&nbsp;    libs=\-lpthread \-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.27\.so
&nbsp;    so=so
&nbsp;    useshrplib=false
&nbsp;    libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.27'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs
&nbsp;    dlext=so
&nbsp;    d\_dlsymun=undef
&nbsp;    ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'
&nbsp;    lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector\-strong'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;:
&nbsp;    HAS\_TIMES
&nbsp;    MYMALLOC
&nbsp;    PERLIO\_LAYERS
&nbsp;    PERL\_COPY\_ON\_WRITE
&nbsp;    PERL\_DONT\_CREATE\_GVSV
&nbsp;    PERL\_MALLOC\_WRAP
&nbsp;    PERL\_OP\_PARENT
&nbsp;    PERL\_PRESERVE\_IVUV
&nbsp;    PERL\_USE\_DEVEL
&nbsp;    USE\_64\_BIT\_ALL
&nbsp;    USE\_64\_BIT\_INT
&nbsp;    USE\_LARGE\_FILES
&nbsp;    USE\_LOCALE
&nbsp;    USE\_LOCALE\_COLLATE
&nbsp;    USE\_LOCALE\_CTYPE
&nbsp;    USE\_LOCALE\_NUMERIC
&nbsp;    USE\_LOCALE\_TIME
&nbsp;    USE\_PERLIO
&nbsp;    USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at May 22 2019 14&#8203;:58&#8203;:04
&nbsp;  %ENV&#8203;:
&nbsp;    PERL2DIR="/home/jkeenan/gitwork/perl2"
&nbsp;    PERLBREW\_HOME="/home/jkeenan/\.perlbrew"
&nbsp;    PERLBREW\_MANPATH="/home/jkeenan/perl5/perlbrew/perls/perl\-5\.28\.0/man"
&nbsp;    PERLBREW\_PATH="/home/jkeenan/perl5/perlbrew/bin&#8203;:/home/jkeenan/perl5/perlbrew/perls/perl\-5\.28\.0/bin"
&nbsp;    PERLBREW\_PERL="perl\-5\.28\.0"
&nbsp;    PERLBREW\_ROOT="/home/jkeenan/perl5/perlbrew"
&nbsp;    PERLBREW\_SHELLRC\_VERSION="0\.84"
&nbsp;    PERLBREW\_VERSION="0\.84"
&nbsp;    PERL\_WORKDIR="/home/jkeenan/gitwork/perl"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    lib
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.31\.0/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.31\.0
&nbsp;    /usr/local/lib/perl5/5\.31\.0/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/5\.31\.0

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat\, 25 May 2019 01&#8203;:56&#8203;:12 GMT\, jkeenan@&#8203;pobox\.com wrote&#8203;:
> pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:
> 
> \#\#\#\#\#
> hostname\(\) doesn't accept any arguments
> 
> The function
> "hostname\(\)" in the Sys&#8203;::Hostname module has always been
> documented to be called with no arguments\. Historically it
> has not enforced this\, and has actually accepted and
> ignored any arguments\. As a result\, some users have got the
> mistaken impression that an argument does something useful\.
> To avoid these bugs\, the function is being made strict\.
> Passing arguments was deprecated in Perl 5\.28\, and will
> become fatal in Perl 5\.32\.
> \#\#\#\#\#
> 
> This entry was recorded in the following commit&#8203;:
> 
> \#\#\#\#\#
> commit 0c9c439d08a65206d442724bcd9fb29fa5a7f937
> Author&#8203;: Zefram \<zefram@&#8203;fysh\.org>
> Date&#8203;:   Wed Dec 6 16&#8203;:28&#8203;:18 2017 \+0000
> 
> document hostname\(@&#8203;args\) deprecation
> 
> Passing arguments to Sys&#8203;::Hostname&#8203;::hostname\(\) is already
> deprecated\. Note the deprecation's agreed removal version
> of Perl 5\.32\. Also document it in perldeprecation\.pod\.
> \#\#\#\#\#
> 
> Make it so\.
> 
> Thank you very much\.
> Jim Keenan

Please review patch attached \(or smoke\-me/jkeenan/rt\-134137\-hostname branch\)\.

Thank you very much\.
\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Prevent-hostname-from-accepting-arguments.patch</summary>

```diff
From 811314492d0ed875a2622bb1f030f711ac5399bb Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Fri, 24 May 2019 22:37:40 -0400
Subject: [PATCH] Prevent hostname() from accepting arguments

Increment $VERSION.

For: RT 134137
---
 ext/Sys-Hostname/Hostname.pm  |  4 ++--
 ext/Sys-Hostname/t/Hostname.t | 17 +++++------------
 2 files changed, 7 insertions(+), 14 deletions(-)

diff --git a/ext/Sys-Hostname/Hostname.pm b/ext/Sys-Hostname/Hostname.pm
index 8b5dde1445..ddf98b8d18 100644
--- a/ext/Sys-Hostname/Hostname.pm
+++ b/ext/Sys-Hostname/Hostname.pm
@@ -16,7 +16,7 @@ use warnings ();
 our $host;
 
 BEGIN {
-    $VERSION = '1.22';
+    $VERSION = '1.23';
     {
 	local $SIG{__DIE__};
 	eval {
@@ -29,7 +29,7 @@ BEGIN {
 
 
 sub hostname {
-  @_ and warnings::warnif("deprecated", "hostname() doesn't accept any arguments. This will become fatal in Perl 5.32");
+  @_ and die("hostname() no longer accepts arguments");
 
   # method 1 - we already know it
   return $host if defined $host;
diff --git a/ext/Sys-Hostname/t/Hostname.t b/ext/Sys-Hostname/t/Hostname.t
index a8c259d7c9..9aaf464339 100644
--- a/ext/Sys-Hostname/t/Hostname.t
+++ b/ext/Sys-Hostname/t/Hostname.t
@@ -10,7 +10,7 @@ BEGIN {
 
 use Sys::Hostname;
 
-use Test::More tests => 4;
+use Test::More tests => 2;
 
 SKIP:
 {
@@ -23,15 +23,8 @@ SKIP:
 }
 
 {
-    use warnings;
-    my $warn;
-    local $SIG{__WARN__} = sub { $warn = "@_" };
-    eval { hostname("dummy") };
-    ok($warn, "warns with an argument");
-    like($warn, qr/hostname\(\) doesn't accept any arguments/,
-         "appropriate message");
-    no warnings "deprecated";
-    undef $warn;
-    eval { hostname("dummy") };
-    is($warn, undef, "no warning when disabled");
+    local $@;
+    eval { hostname("dummy"); };
+    like($@, qr/hostname\(\) no longer accepts arguments/,
+        "hostname no longer accepts arguments");
 }
-- 
2.17.1


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri\, 24 May 2019 19&#8203;:48&#8203;:35 \-0700\, jkeenan wrote&#8203;:
> On Sat\, 25 May 2019 01&#8203;:56&#8203;:12 GMT\, jkeenan@&#8203;pobox\.com wrote&#8203;:
> > pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:
> >
> > \#\#\#\#\#
> > hostname\(\) doesn't accept any arguments
> >
> > The function
> > "hostname\(\)" in the Sys&#8203;::Hostname module has always been
> > documented to be called with no arguments\. Historically it
> > has not enforced this\, and has actually accepted and
> > ignored any arguments\. As a result\, some users have got the
> > mistaken impression that an argument does something useful\.
> > To avoid these bugs\, the function is being made strict\.
> > Passing arguments was deprecated in Perl 5\.28\, and will
> > become fatal in Perl 5\.32\.
> > \#\#\#\#\#
> >
> > This entry was recorded in the following commit&#8203;:
> >
> > \#\#\#\#\#
> > commit 0c9c439d08a65206d442724bcd9fb29fa5a7f937
> > Author&#8203;: Zefram \<zefram@&#8203;fysh\.org>
> > Date&#8203;:   Wed Dec 6 16&#8203;:28&#8203;:18 2017 \+0000
> >
> > document hostname\(@&#8203;args\) deprecation
> >
> > Passing arguments to Sys&#8203;::Hostname&#8203;::hostname\(\) is already
> > deprecated\. Note the deprecation's agreed removal version
> > of Perl 5\.32\. Also document it in perldeprecation\.pod\.
> > \#\#\#\#\#
> >
> > Make it so\.
> >
> > Thank you very much\.
> > Jim Keenan
> 
> Please review patch attached \(or smoke\-me/jkeenan/rt\-134137\-hostname
> branch\)\.
> 
> Thank you very much\.

\- @&#8203;\_ and warnings&#8203;::warnif\("deprecated"\, "hostname\(\) doesn't accept any arguments\. This will become fatal in Perl 5\.32"\);
\+ @&#8203;\_ and die\("hostname\(\) no longer accepts arguments"\); 

Two issues&#8203;:

\- this should probably use croak\(\) instead of die\(\) so the error is produced at the caller

\- Richard's point about the error message in 

https://www.nntp.perl.org/group/perl.perl5.porters/2019/05/msg254903.html

which didn't make it to the ticket\.

Tony

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue\, 04 Jun 2019 23&#8203;:53&#8203;:53 GMT\, tonyc wrote&#8203;:
> On Fri\, 24 May 2019 19&#8203;:48&#8203;:35 \-0700\, jkeenan wrote&#8203;:
> > On Sat\, 25 May 2019 01&#8203;:56&#8203;:12 GMT\, jkeenan@&#8203;pobox\.com wrote&#8203;:
> > > pod/perldeprecation\.pod contains this entry for perl\-5\.32&#8203;:
> > >
> > > \#\#\#\#\#
> > > hostname\(\) doesn't accept any arguments
> > >
> > > The function
> > > "hostname\(\)" in the Sys&#8203;::Hostname module has always been
> > > documented to be called with no arguments\. Historically it
> > > has not enforced this\, and has actually accepted and
> > > ignored any arguments\. As a result\, some users have got the
> > > mistaken impression that an argument does something useful\.
> > > To avoid these bugs\, the function is being made strict\.
> > > Passing arguments was deprecated in Perl 5\.28\, and will
> > > become fatal in Perl 5\.32\.
> > > \#\#\#\#\#
> > >
> > > This entry was recorded in the following commit&#8203;:
> > >
> > > \#\#\#\#\#
> > > commit 0c9c439d08a65206d442724bcd9fb29fa5a7f937
> > > Author&#8203;: Zefram \<zefram@&#8203;fysh\.org>
> > > Date&#8203;:   Wed Dec 6 16&#8203;:28&#8203;:18 2017 \+0000
> > >
> > > document hostname\(@&#8203;args\) deprecation
> > >
> > > Passing arguments to Sys&#8203;::Hostname&#8203;::hostname\(\) is already
> > > deprecated\. Note the deprecation's agreed removal version
> > > of Perl 5\.32\. Also document it in perldeprecation\.pod\.
> > > \#\#\#\#\#
> > >
> > > Make it so\.
> > >
> > > Thank you very much\.
> > > Jim Keenan
> >
> > Please review patch attached \(or smoke\-me/jkeenan/rt\-134137\-hostname
> > branch\)\.
> >
> > Thank you very much\.
> 
> \- @&#8203;\_ and warnings&#8203;::warnif\("deprecated"\, "hostname\(\) doesn't accept any
> arguments\. This will become fatal in Perl 5\.32"\);
>  \+ @&#8203;\_ and die\("hostname\(\) no longer accepts arguments"\);
> 
> Two issues&#8203;:
> 
> \- this should probably use croak\(\) instead of die\(\) so the error is
> produced at the caller
> 
> \- Richard's point about the error message in
> 
> https://www.nntp.perl.org/group/perl.perl5.porters/2019/05/msg254903.html
> 
> which didn't make it to the ticket\.
> 
> Tony

Thanks for the code review\.  Incorporated suggestions and merged to blead in commit 82b53111b29b7b8060044482d261a31ccd5ac001\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'pending release'
